### PR TITLE
Split into multiple files.  

### PR DESCRIPTION
**Pull request by paris3200**
I've very new to contributing to someone else's project, however I thought I'd try to tackle #6.  Please let me know if I need to fix anything.  

Thanks.

### CODE REVIEW

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Thanks! I think it'd be great to solve #1 (at least the `setup.py` part) at the same time, though. Otherwise, it's not really possible anymore to run it easily, because it's now not just a single file which can be run anymore.

Would you like to take a look at that? Then, things like `run` and `main` from `output.py` should be in a `main.py` or so, and the `setup.py` should have a `console_scripts` entry point for it.

Also, I guess it'd be good to switch to the `src` packaging layout while you're at it as well.

Some related docs:

- [Packaging Python Projects — Python Packaging User Guide](https://packaging.python.org/tutorials/packaging-projects/)
- [Packaging a python library | ionel's codelog](https://blog.ionelmc.ro/2014/05/25/python-packaging/) (the part about the `src` layout)

**Comment by paris3200** [comment]
Thank you for supplying the related docs.  I'll take care of it moving to a `src` packaging and creating the `setup.py`.

**Comment by paris3200** [comment]
I believe I've gotten the bugs worked out now.  Ideally there would be more tests, but I'm still trying to understand the code. I believe for tests the API results needs to be mocked.

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Thanks for the quick update! This is looking great except of a couple of minor things I've commented on.

Also I personally prefer doing e.g. `from cloclify import output` and using `output.dump` (same for all other modules, i.e. importing modules instead of individual functions/classes). This makes it clearer to see where the code is coming from.

For this (relatively small) project I guess it's fine to have the more fine-grained imports, just something to think about. Feel free to pick whatever you prefer.

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        "Programming Language :: Python :: 3",
+        "License :: OSI Approved :: MIT License",
+        "Operating System :: OS Independent",
+    ],
+    package_dir={"": "src"},
+    packages=setuptools.find_packages(where="src"),
+    python_requires=">=3.6",
+
+    entry_points={
+         'console_scripts': [
```
[the remark this answers is not in the document]
Comment (paris3200) [reply setup.py]: That's what I get for not using pre-commit.  I don't consider that nitpicking at all.  I appreciate you enforcing standards.

**Comment by paris3200** [comment]
Thank you for reviewing this so quickly and thoroughly.  I really appreciate you explaining the "why" something needs to be changed.  I've made the requested changes.  I went ahead and switched to the importing of the entire module instead of the function/class.  You're right, it does make things more clear.  However, this did required me to rename the local variables.  These changes are in 055eb91.

**Comment by The-Compiler** [comment]
I just tried installing this in a clean [virtualenv](https://virtualenv.pypa.io/en/latest/index.html) via:

```
$ python3 -m venv .c
$ .c/bin/pip install git+https://github.com/paris3200/cloclify.git
$ .c/bin/cloclify
```

and got:

```pytb
Traceback (most recent call last):
  File "/home/florian/tmp/.c/bin/cloclify", line 33, in <module>
    sys.exit(load_entry_point('cloclify==0.0.1', 'console_scripts', 'cloclify')())
  File "/home/florian/tmp/.c/bin/cloclify", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/usr/lib/python3.9/importlib/metadata.py", line 77, in load
    module = import_module(match.group('module'))
  File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 790, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/home/florian/tmp/.c/lib/python3.9/site-packages/cloclify/main.py", line 4, in <module>
    import rich.console
ModuleNotFoundError: No module named 'rich'
```

It looks like the [`install_requires` key](https://setuptools.readthedocs.io/en/latest/userguide/dependency_management.html#declaring-dependencies) is still missing in `setup.py`. The packages which should be listed there are essentially the same as in `requirements.txt` - see [setup.py vs. requirements.txt](https://caremad.io/posts/2013/07/setup-vs-requirement/) for details on why the duplication exists.

By the way, if I suddenly stop responding here for more than week or so, feel free to ping me again - there's a lot going on at the moment, and most contributions I get are in another project ([qutebrowser](https://www.qutebrowser.org/)) - so chances are a notification mail got lost and I forgot about this for a bit :slight_smile:.

> I really appreciate you explaining the "why" something needs to be changed.

Glad to hear! I teach things (including Python) to people as a day-job, so I'm happy to explain stuff - and it's great to see that someone is showing cloclify some love, as I'm mostly busy with qutebrowser things already. :+1:

**Comment by paris3200** [comment]
I've been a qutebrowser user since the early years.  There is no way you'd remember, but you actually helped me with a few issues I had with it on IRC.  It's a great browser that has really changed my workflow for the better.  

I updated the setup.py.  I ran it in a virtualenv and it works.  Going forward, would it be possible to setup a CI of some kind?  Once this is done I think I'll try tackling some tests.  I'd like to add some features, but would really prefer to have some basic testing in place before doing that.

**Comment by The-Compiler** [comment]
I had to check my old logs, and I actually remember this one kinda (though I didn't remember who it was):

```
2015-04-01 17:46:02 Paris   qutebrowser is now showing all webpages titled slightly. 
2015-04-01 17:46:25 Paris   Is this a april fools joke?
```

:wink:

---

Looks great and seems to work now, thanks again! :+1: 

I agree on setting up a CI, ideally [GitHub Actions](https://docs.github.com/en/actions). It's enabled for this repository, so if you want to contribute a PR adding a suitable workflow, that'd be most appreciated! If not, I can take a look, but probably not in the next 2-3 weeks.

You might be able to adopt something like this:

- https://github.com/The-Compiler/pytest-xvfb/blob/master/.github/workflows/main.yml
- https://github.com/The-Compiler/pytest-xvfb/blob/master/tox.ini

(except with some updated Python versions I guess)

**Comment by paris3200** [comment]
You had to did way back in the logs for that.  That's hilarious.  

Glad we were able to work out the issues and get this merged.  I really appreciate you working with me on this.  I'll see about setting up GitHub Actions.  I'm pretty busy at the moment with other commitments, but I'll try to carve out some time.  

PS:  Shouldn't this close #6?

**Pull request merged**